runtime/shared_common/CacheMap.cpp

```diff
@@ -7012,45 +7009,153 @@ SH_CacheMap::startupForStats(J9VMThread* currentThread, SH_OSCache * oscache, U_
+					retval = CC_STARTUP_FAILED;
+					break;
+				}
+				tempInfo.javacoreData.freeBytes = ccToUse->getFreeAvailableBytes();
+				tempInfo.isJavaCorePopulated = 1;
+				SH_OSCache_Info* newElement = (SH_OSCache_Info*) pool_newElement(*lowerLayerList);
+				if (NULL == newElement) {
+					retval = CC_STARTUP_FAILED;
+					break;
+				}
+				memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
 			} else {
 				retval = CC_STARTUP_FAILED;
+				break;
 			}
-			goto done;
 		}
-		cache = cache->getNext();
-	}
-#endif
+		ccToUse = ccToUse->getPrevious();
+	} while (NULL != ccToUse && CC_STARTUP_OK == retval);
 
 done:
 	if (retval != CC_STARTUP_OK) {
+		if (CC_STARTUP_FAILED == retval) {
+			if (NULL != *lowerLayerList) {
+				pool_kill(*lowerLayerList);
+			}
+		}
 		shutdownForStats(currentThread);
 	}
 	return retval;
 }
 
+/**
+ * Start up the lower layer shared cache for statistics.
+ *
+ * @param [in] currentThread Pointer to J9VMThread structure for the current thread
+ * @param [in] oscache an exiting top layer SH_OSCache
+ * @param [out] lowerLayerStatsList A list of SH_OSCache_Info for all lower layer caches.
+ *
+ * @retval CC_STARTUP_OK (0) success
+ * @retval CC_STARTUP_FAILED(-1) or CC_STARTUP_CORRUPT (-2) on failure.
+ */
+IDATA
+SH_CacheMap::startupLowerLayerForStats(J9VMThread* currentThread, const char* ctrlDirName, UDATA groupPerm, SH_OSCache *oscache, J9Pool** lowerLayerList)
+{
+	IDATA rc = CC_STARTUP_OK;
+	SH_CompositeCacheImpl* ccToUse = _ccHead;
+	const char* cacheName = NULL;
+	U_32 cacheType = oscache->getCacheType();
+	char cacheUniqueID[J9SHR_UNIQUE_CACHE_ID_BUFSIZE];
+	J9JavaVM *vm = currentThread->javaVM;
+	char cacheDirBuf[J9SH_MAXPATH];
+	SH_OSCache::getCacheDir(vm, ctrlDirName, cacheDirBuf, J9SH_MAXPATH, cacheType, false);
+	
+	do {
+		const char* cacheUniqueIDPtr = NULL;
+		UDATA idLen = 0;
+		
+		IDATA preqRC = getPrereqCache(currentThread, cacheDirBuf, ccToUse, true, &cacheUniqueIDPtr, &idLen);
+		I_8 layer = 0;
+
+		if (0 > preqRC) {
+			if (CM_CACHE_CORRUPT == preqRC) {
+				rc = CC_STARTUP_CORRUPT;
+				SH_Managers::ManagerWalkState state;
+				SH_Manager* walkManager = managers()->startDo(currentThread, 0, &state);
+				while (walkManager) {
+					/* Corruption on the metadata detected. We will return from this function. Clean up the managers before that. */ 
+					walkManager->cleanup(currentThread);
+					walkManager = managers()->nextDo(&state);
+				}
+			} else {
+				rc = CC_STARTUP_FAILED;
+			}
+			break;
+		} else if (1 == preqRC) {
+			PORT_ACCESS_FROM_VMC(currentThread);
+			UDATA reqBytes = SH_CompositeCacheImpl::getRequiredConstrBytesWithCommonInfo(false, true);
+			SH_CompositeCacheImpl* allocPtr = (SH_CompositeCacheImpl*)j9mem_allocate_memory(reqBytes, J9MEM_CATEGORY_CLASSES);
+			if (NULL == allocPtr) {
+				rc = CC_STARTUP_FAILED;
+				break;
+			}
+			char cacheNameBuf[USER_SPECIFIED_CACHE_NAME_MAXLEN];
+			Trc_SHR_Assert_True((idLen + 1) <= sizeof(cacheUniqueID));
```

Comment:
`idLen < sizeof(cacheUniqueID)` would also work.

Reply:
Changed to use idLen < sizeof(cacheUniqueID)